extensions/amp-consent/0.1/consent-policy-manager.js

```diff
@@ -144,18 +157,26 @@ export class ConsentPolicyInstance {
   consentStateChangeHandler(consentId, state) {
     // TODO: Keeping an array can have performance issue, change to using a map
     // if necessary.
+    dev().assert(this.itemMap_[consentId] != undefined,
+        `cannot find ${consentId} in policy state`);
     if (state == CONSENT_ITEM_STATE.GRANTED) {
-      const index = this.pendingItems_.indexOf(consentId);
-      if (index > -1) {
-        this.pendingItems_.splice(index, 1);
+      if (this.itemMap_[consentId] == CONSENT_ITEM_STATE.UNKNOWN) {
+        this.pendingItemCount_--;
+      }
+      if (this.itemMap_[consentId] == CONSENT_ITEM_STATE.REJECTED) {
+        this.rejectedItemCount_--;
       }
+      this.itemMap_[consentId] = CONSENT_ITEM_STATE.GRANTED;
     }
 
     if (state == CONSENT_ITEM_STATE.REJECTED) {
-      const index = this.pendingItems_.indexOf(consentId);
-      if (index == -1) {
-        this.pendingItems_.push(consentId);
+      if (this.itemMap_[consentId] == CONSENT_ITEM_STATE.UNKNOWN) {
+        this.pendingItemCount_--;
       }
+      if (this.itemMap_[consentId] != CONSENT_ITEM_STATE.REJECTED) {
+        this.rejectedItemCount_++;
+      }
+      this.itemMap_[consentId] = CONSENT_ITEM_STATE.REJECTED;
```

Comment:
Minor:

```js
const granted = (state == CONSENT_ITEM_STATE.GRANTED);
const rejected = (state == CONSENT_ITEM_STATE.REJECTED);

if (!granted && !rejected) {
  return;
}

switch (this.itemMap_[consentId]) {
  case CONSENT_ITEM_STATE.UNKNOWN: 
    this.pendingItemCount_--; 
    break;
  case CONSENT_ITEM_STATE.REJECTED: 
    this.rejectedItemCount_ += (granted ? 1 : -1); 
    break;
}
this.itemMap_[consentId] = state;
```

Reply:
Actually I'd recommend removing the `pendingItemCount_` and `rejectedItemCount_` ivars and just reducing over `Object.keys(itemMap_)` in `evaluate_()`. State tracking like this is a source of bugs and harder to read.

Reply:
I can't change to this now. Right now, the `consentStateManager` will call this `consentStateChangeHandler` whenever there is an user action. This also includes situations like user grant consent, then grant same consent again from other UI flow (Note: it is required to allow user to change their given consent state afterwards).

I agree this is hard to read and a source of bugs : ( I can leave as it is and then migrate to what you suggest here after adding `hasConsentStateReallyChanged` check to `consentStateManager`.

And to `just reducing over Object.keys(itemMap_) in evaluate_()`? I am a bit concerned about performance in that case. Object.keys requires traversing the map everytime? Would that be OK?

Reply:
Probably ok unless `itemMap_` is very large. Premature optimization something something. :)

Reply:
😄 Yeah I assume itemMap_ to be fairly small. Switched.